lib/models/room.js

```diff
@@ -237,43 +316,226 @@ Room.prototype.getAvatarUrl = function(baseUrl, width, height, resizeMethod,
+ *
+ * @param {string=} paginationToken   token for the next batch of events
+ *
  * @fires module:client~MatrixClient#event:"Room.timeline"
+ *
  */
-Room.prototype.addEventsToTimeline = function(events, toStartOfTimeline) {
-    var stateContext = toStartOfTimeline ? this.oldState : this.currentState;
+Room.prototype.addEventsToTimeline = function(events, toStartOfTimeline,
+                                              timeline, paginationToken) {
+    if (!timeline) {
+        timeline = this._liveTimeline;
+    }
 
-    function checkForRedaction(redactEvent) {
-        return function(e) {
-            return e.getId() === redactEvent.event.redacts;
-        };
+    if (!toStartOfTimeline && timeline == this._liveTimeline) {
+        // special treatment for live events
+        this._addLiveEvents(events);
+        return;
+    }
+
+    // we need to handle the following case:
+    //
+    // We already know about two timelines. One of these contains just
+    // event M, and one contains just event P:
+    //
+    //    timeline1      timeline2
+    //       [M]            [P]
+    //
+    // We are now processing a new pagination response from the server.
+    // This contains the following events: [M, N, P, R, S].
+    //
+    // 1. First, we ignore event M, since we already know about it.
+    //
+    // 2. Next, we append N to timeline 1.
+    //
+    // 3. Next, we don't add event P, since we already know about it,
+    //    but we do link togehter the timelines. We now have:
+    //
+    //    timeline1      timeline2
+    //     [M, N] <----> [P]
+    //
+    // 4. We now need to append R and S to timeline2. So finally, we have:
+    //
+    //    timeline1      timeline2
+    //     [M, N] <----> [P, R, S]
+    //
+    // The important thing to note in the above is that we switched the
+    // timeline we were adding events to, after we found a neighbour.
+
+    // XXX: TODO: but timeline2 might already be joined to timeline3: we need
+    // to skip all the way to the end of the linked list!
+
+    var updateToken = false;
+    for (var i = 0; i < events.length; i++) {
+        var existingTimeline = this._checkForNewEventInExistingTimeline(
+            events[i], timeline, toStartOfTimeline);
+
+        if (existingTimeline) {
+            timeline = existingTimeline;
+            updateToken = false;
+        } else {
+            this._addEventToTimeline(events[i], timeline, toStartOfTimeline);
+            updateToken = true;
+        }
+    }
+    if (updateToken) {
+        timeline.setPaginationToken(paginationToken, toStartOfTimeline);
+    }
+};
+
+/**
+ * Check if this event is already in a timeline, and join up the timelines if
+ * necessary
+ *
+ * @param {MatrixEvent} event           event to add
+ * @param {EventTimeline} timeline      timeline we think we should add to
+ * @param {boolean} toStartOfTimeline   true if we're adding to the start of
```

Comment:
This determines which end of the timeline we *may* be gluing to another timeline. It might be better to make this a bit clearer